Add multiple airspaces to XCSoar

Eldalie opened the pull request:
Rebase of https://github.com/XCSoar/XCSoar/pull/471/files

Closes https://github.com/XCSoar/XCSoar/issues/425

I read the last comment about the UI and other suggestions, but right now I'm more annoyed by the change in airspace selection during flight, rather than the fact that it takes two steps to select from the list of airspaces on the ground.
This can be improved later.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added support for selecting and managing multiple airspace files, allowing users to configure and load several airspace data files simultaneously.
  * Introduced a new multi-file picker dialog for user-friendly selection and management of multiple files in configuration panels.
  * Added a new data field type to support multiple file selections in forms and dialogs.

* **Improvements**
  * Updated configuration panels and documentation to reflect the new multi-file airspace selection.
  * Enhanced profile handling to store, migrate, and retrieve multiple airspace file paths.
  * Improved file selection widgets to support and display multiple files.
  * Streamlined airspace file loading across application startup and testing utilities for multi-file support.

* **Bug Fixes**
  * Ensured consistent handling of multiple airspace files across application startup, configuration, and test utilities.

* **Documentation**
  * Updated user manuals to describe the new multi-file airspace configuration process.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

lordfolken commented on src/Form/DataField/MultiFile.hpp:
```diff
@@ -0,0 +1,116 @@
+  void Dec() noexcept override {}
+};
+
+#endif
```
Please fix the newline. (Also please ensure your editor is adding them.)

lordfolken commented on src/Form/DataField/MultiFile.cpp:
```diff
@@ -0,0 +1,172 @@
+/*
+Copyright_License {
+  XCSoar Glide Computer - http://www.xcsoar.org/
+  Copyright (C) 2000-2016 The XCSoar Project
+  A detailed list of copyright holders can be found in the file "AUTHORS".
+  This program is free software; you can redistribute it and/or
+  modify it under the terms of the GNU General Public License
+  as published by the Free Software Foundation; either version 2
+  of the License, or (at your option) any later version.
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+}
```
Please replace copyright notices as they are in the other files, so we don't need to repeat this text endlessly.

lordfolken commented on src/Profile/PathValue.cpp:
```diff
@@ -28,6 +31,43 @@ ProfileMap::GetPath(std::string_view key) const noexcept
+    if (i.empty())
+      continue;
+
+#ifdef _UNICODE
+    std::wstring file_string(i);
+#else
+    std::string file_string(i);
+#endif
```
We have a tstring macro. Please use this instead.

lordfolken commented on src/Profile/PathValue.cpp:
```diff
@@ -28,6 +31,43 @@ ProfileMap::GetPath(std::string_view key) const noexcept
+
+    if (!path.EndsWithIgnoreCase(_(".txt")) && path.EndsWithIgnoreCase(_(".air")))
+      continue;
+
+    paths.push_back(ExpandLocalPath(AllocatedPath(path)));
+
+  }
+
+
+  return paths;
+}
+
```
Please remove the excess newlines

lordfolken commented on src/Form/DataField/MultiFile.cpp:
```diff
@@ -0,0 +1,172 @@
+
+    display_string +=
+        file_datafield.GetItem(file_datafield.Find(path)).filename.c_str();
+    display_string += _(" ");
```
This will add an empty space at the end.

lordfolken reviewed (changes requested):
So first of all Thank you. :)

Some general things:
 * no uncessary stuff please
 * please format your code properly (lots of unnecessary random newlines)
 * please fix your added files and add a newline at the end
 * use tstring macro where possible
 * copyright headers like other files please

some other comments inline.

Eldalie commented on src/Profile/PathValue.cpp:
```diff
@@ -28,6 +31,43 @@ ProfileMap::GetPath(std::string_view key) const noexcept
+    Path path(file_string.c_str());
+
+    if (!path.EndsWithIgnoreCase(_(".txt")) && path.EndsWithIgnoreCase(_(".air")))
+      continue;
```
Yes, I forgot the '!' but your version is better.

lordfolken commented:
At the moment it fails to compile: 

```
 src/Dialogs/DataField.cpp: In function 'bool EditDataFieldDialog(const TCHAR*, DataField&, const TCHAR*)':
src/Dialogs/DataField.cpp:36:71: error: no matching function for call to 'MultiFileDataField::MultiFileDataField(DataField&)'
   36 |     return MultiFilePicker(caption, static_cast<MultiFileDataField>(df),
      |                                                                       ^
In file included from src/Dialogs/DataField.cpp:15:
./src/Form/DataField/MultiFile.hpp:34:12: note: candidate: 'MultiFileDataField::MultiFileDataField(DataFieldListener*)'
   34 |   explicit MultiFileDataField(DataFieldListener *listener = nullptr);
      |            ^~~~~~~~~~~~~~~~~~
./src/Form/DataField/MultiFile.hpp:34:50: note:   no known conversion for argument 1 from 'DataField' to 'DataFieldListener*'
   34 |   explicit MultiFileDataField(DataFieldListener *listener = nullptr);
      |                               ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
./src/Form/DataField/MultiFile.hpp:19:7: note: candidate: 'MultiFileDataField::MultiFileDataField(MultiFileDataField&&)'
   19 | class MultiFileDataField final : public DataField {
      |       ^~~~~~~~~~~~~~~~~~
./src/Form/DataField/MultiFile.hpp:19:7: note:   no known conversion for argument 1 from 'DataField' to 'MultiFileDataField&&'
make[2]: *** [build/compile.mk:104: output/DEBIAN-UNIX/opt/src/Dialogs/DataField.o] Error 1
```

Eldalie commented:
Sorry, I haven’t requested a new review as I haven’t had time to finish resolving all your issues.
I’ve moved the PR to draft.

lordfolken commented on src/Dialogs/DataField.cpp:
```diff
@@ -43,18 +48,15 @@ EditDataFieldDialog(const TCHAR *caption, DataField &df,
     return true;
   } else if (type == DataField::Type::DATE) {
     auto &dfd = static_cast<DataFieldDate &>(df);
     BrokenDate date = dfd.GetValue();
-    if (!DateEntryDialog(caption, date, true))
-      return false;
+    if (!DateEntryDialog(caption, date, true)) return false;
```
no functional change. I find the newline return false better to read.

lordfolken commented on src/Dialogs/DataField.cpp:
```diff
@@ -64,21 +66,19 @@ EditDataFieldDialog(const TCHAR *caption, DataField &df,
       /* with signed range has to avoid the length of negative AND
-      * positiv numbers */
-      int value = dfi.GetValue();  // min is < 0!
+       * positiv numbers */
```
If fixing the indent, can you also fix the spelling mistake? 

The two spaces after function name in former line 76 are intentional.

lordfolken commented on src/Dialogs/DataField.cpp:
```diff
@@ -102,18 +102,16 @@ EditDataFieldDialog(const TCHAR *caption, DataField &df,
-                                     caption, value, type);
-      if (!new_value)
-        return false;
+      auto new_value =
+          dlg.ShowModal(Java::GetEnv(), *context, caption, value, type);
+      if (!new_value) return false;
```
Stuff like this makes it hard to review the changes, as there is no functional change.

lordfolken commented on src/Dialogs/MultiFilePicker.cpp:
```diff
@@ -0,0 +1,122 @@
+  virtual void OnPaintItem(Canvas &canvas, const PixelRect rc,
+                           unsigned i) noexcept override
+  {
+
```
remove space

lordfolken commented on src/Dialogs/MultiFilePicker.cpp:
```diff
@@ -0,0 +1,122 @@
+  std::vector<Path> &active_files;
+
+public:
+  explicit MultiFilePickerSupport(std::vector<Path> &_active_files)
+      : active_files(_active_files)
+  {
+  }
```
I know { } on new line. but since this construtctor is empty you can put them on the same line as the definiton.

lordfolken commented on src/Dialogs/MultiFilePicker.cpp:
```diff
@@ -0,0 +1,122 @@
+  int result = dialog.ShowModal();
+
+  if (result == mrRemove) {
+
```
remove space

lordfolken commented on src/Dialogs/MultiFilePicker.cpp:
```diff
@@ -0,0 +1,122 @@
+  dialog.AddButton(_("Cancel"), mrCancel);
+
+  dialog.EnableCursorSelection();
+
```
remove newlines

lordfolken commented on src/Dialogs/Settings/Panels/SiteConfigPanel.cpp:
```diff
@@ -31,69 +30,72 @@ class SiteConfigPanel final : public RowFormWidget {
 public:
-  SiteConfigPanel()
-    :RowFormWidget(UIGlobals::GetDialogLook()) {}
+  SiteConfigPanel() : RowFormWidget(UIGlobals::GetDialogLook()) {}
```
Please keep on same line. (I know 80 character limit, but it isn´t a hard limit if it enhances readability.)

lordfolken commented on src/Dialogs/Settings/Panels/SiteConfigPanel.cpp:
```diff
@@ -31,69 +30,72 @@ class SiteConfigPanel final : public RowFormWidget {
+          ProfileKeys::AirfieldFile, _T("*.txt\0"), FileType::WAYPOINTDETAILS);
+
+  AddMultipleFiles(_("Airspace Files"),
+                   _("List of active airspace files. Use the Add and Remove "
+                     "buttons to activate or deactivate "
+                     " airspace files respectively. Supported file types are: "
+                     "Openair (.txt /.air), and Tim Newport-Pearce (.sua). "),
```
Please don't try to format the text here. That is really a task of the AddMultipleFiles function.

Eldalie replied:
Idem: I formatted my change using clang-format, and this follows the current formatting rules. So either we update the formatting rules in a separate MR, or I keep this change as is.

The formatting applied here is only in the C++ file, not in the UI.

lordfolken replied:
You are indeed correct that that is the formatting of clang. Please however remove the resulting double space between deactivate and airspace.

lordfolken commented on src/Dialogs/Settings/Panels/SiteConfigPanel.cpp:
```diff
@@ -31,69 +30,72 @@ class SiteConfigPanel final : public RowFormWidget {
+                   FileType::AIRSPACE);
 
   AddFile(_("FLARM Device Database"),
-          _("The name of the file containing information about registered FLARM devices."),
-          ProfileKeys::FlarmFile, _T("*.fln\0"),
-          FileType::FLARMNET);
+          _("The name of the file containing information about registered "
+            "FLARM devices."),
+          ProfileKeys::FlarmFile, _T("*.fln\0"), FileType::FLARMNET);
```
non functional change

lordfolken commented on src/Dialogs/Settings/Panels/SiteConfigPanel.cpp:
```diff
@@ -105,22 +107,25 @@ SiteConfigPanel::Save(bool &_changed) noexcept
-  WaypointFileChanged |= SaveValueFileReader(WaypointFile, ProfileKeys::WaypointFile);
-  WaypointFileChanged |= SaveValueFileReader(AdditionalWaypointFile, ProfileKeys::AdditionalWaypointFile);
-  WaypointFileChanged |= SaveValueFileReader(WatchedWaypointFile, ProfileKeys::WatchedWaypointFile);
+  WaypointFileChanged |=
+      SaveValueFileReader(WaypointFile, ProfileKeys::WaypointFile);
+  WaypointFileChanged |= SaveValueFileReader(
+      AdditionalWaypointFile, ProfileKeys::AdditionalWaypointFile);
+  WaypointFileChanged |= SaveValueFileReader(WatchedWaypointFile,
+                                             ProfileKeys::WatchedWaypointFile);
 
-  AirspaceFileChanged = SaveValueFileReader(AirspaceFile, ProfileKeys::AirspaceFile);
-  AirspaceFileChanged |= SaveValueFileReader(AdditionalAirspaceFile, ProfileKeys::AdditionalAirspaceFile);
+  AirspaceFileChanged |= SaveValueMultiFileReader(
+      AirspaceFileList, ProfileKeys::AirspaceFileList);
```
the 80 char limit isnt a hard limit if it enables readability. you can keep it on the same line.

lordfolken commented on src/Form/DataField/File.cpp:
```diff
@@ -22,37 +22,35 @@
-    _T("user.cup"),
-    nullptr
-  };
+  static const TCHAR *const ifiles[] = {_T("xcsoar-checklist.txt"),
+                                        _T("xcsoar-flarm.txt"),
+                                        _T("xcsoar-marks.txt"),
+                                        _T("xcsoar-persist.log"),
+                                        _T("xcsoar-startup.log"),
+                                        _T("xcsoar.log"),
+                                        _T("xcsoar-rasp.dat"),
+                                        _T("user.cup"),
+                                        nullptr};
```
I think it was ok before.

Eldalie commented:
Thanks, I’ll undo the formatting and split it into multiple commits.

But honestly, clang-format should just handle all the spacing, indentation, and formatting. It's such a pain when different files have different styles.

lordfolken commented:
Any improvements you can suggest to https://github.com/XCSoar/XCSoar/blob/master/.clang-format please do. 

I agree that that is a pain, and i would like for any changes to be in the defined .clang-format. But you have to see that this is one of the oldest code bases i know at over 23 years old, and a lot has changed.

lordfolken commented:
Thank you for putting the PR into shape. Much more readable. 

A few points:
UI:
 * If we rename the first dialog to "Selected Airspace Files" it might become a bit clearer. 
 * In the filepicker "Add file"  there is an empty line. Selecting this one and hittng add witll crash xcsoar. 

UpgradePath:
 * I see that the old variables are no longer addressed. We should maybe read them out and write the new values in, otherwise all people's profiles will be broken.

lordfolken commented:
I selected 3 files, and moved them out of the ~/.xcsoar directory. Starting XCSoar throws errors about the missing files, then crashes when trying to open config->system:
xcsoar: ./src/util/TrivialArray.hxx:140: constexpr TrivialArray<T, max>::const_reference TrivialArray<T, max>::operator[](size_type) const [with T = FileDataField::Item; long unsigned int max = 512; const_reference = const FileDataField::Item&; size_type = long unsigned int]: Assertion `i < size()' failed.
Aborted

Eldalie commented:
Thanks for your review! I hope I’ll get this work done one day.

My last flight was rough—I had to switch between France, Italy, and Switzerland all in-flight !

I’ve addressed all your points.
I tried fixing the tests, but I’m not really sure how to run them.
I also added an upgrade from the old profile in
src/Profile/File.cpp,
but I’m not sure if that’s actually the best way to do it.

lordfolken commented:
@coderabbitai full review

Eldalie commented:
@coderabbitai Any idea why the build fail on windows with this message and how to correct it ?
`/usr/bin/i686-w64-mingw32-ld: output/PC/dbg/profile.a(PathValue.o): in function `ZNK10ProfileMap16GetMultiplePathsESt17basic_string_viewIcSt11char_traitsIcEE':
/__w/XCSoar/XCSoar/src/Profile/PathValue.cpp:54: undefined reference to `gettext(wchar_t const*)'
/usr/bin/i686-w64-mingw32-ld: /__w/XCSoar/XCSoar/src/Profile/PathValue.cpp:55: undefined reference to `gettext(wchar_t const*)'
/usr/bin/i686-w64-mingw32-ld: /__w/XCSoar/XCSoar/src/Profile/PathValue.cpp:56: undefined reference to `gettext(wchar_t const*)'
i686-w64-mingw32-g++ -Og -funit-at-a-time -ffast-math -g -Wl,--major-subsystem-version=5 -Wl,--minor-subsystem-version=00 -Wl,-subsystem,console -static-libstdc++ -static-libgcc -L./output/PC/lib/i686-w64-mingw32/lib     -march=i586 -mwin32 -mwindows -mms-bitfields   -o output/PC/bin/ReadProfileInt.exe output/PC/dbg/src/LocalPath.o output/PC/dbg/src/Profile/Profile.o output/PC/dbg/test/src/FakeLogFile.o output/PC/dbg/test/src/ReadProfileInt.o output/PC/dbg/profile.a output/PC/dbg/io.a output/PC/dbg/fmt.a output/PC/dbg/libos.a output/PC/dbg/util.a   -lz -L/__w/XCSoar/XCSoar/output/PC/lib/i686-w64-mingw32/lib -lfmt  -lws2_32  -lwinmm   
collect2: error: ld returned 1 exit status
make: *** [build/test.mk:1040: output/PC/bin/ReadProfileString.exe] Error 1
make: *** Waiting for unfinished jobs....
/usr/bin/i686-w64-mingw32-ld: output/PC/dbg/profile.a(PathValue.o): in function `ZNK10ProfileMap16GetMultiplePathsESt17basic_string_viewIcSt11char_traitsIcEE':
/__w/XCSoar/XCSoar/src/Profile/PathValue.cpp:54: undefined reference to `gettext(wchar_t const*)'
/usr/bin/i686-w64-mingw32-ld: /__w/XCSoar/XCSoar/src/Profile/PathValue.cpp:55: undefined reference to `gettext(wchar_t const*)'
/usr/bin/i686-w64-mingw32-ld: /__w/XCSoar/XCSoar/src/Profile/PathValue.cpp:56: undefined reference to `gettext(wchar_t const*)'`

Eldalie commented:
@lordfolken What is missing to merge this  ?

lordfolken commented on src/Profile/PathValue.cpp:
```diff
@@ -28,6 +32,36 @@ ProfileMap::GetPath(std::string_view key) const noexcept
+
+    Path path(file_string.c_str());
+
+    if (!(path.EndsWithIgnoreCase(_T(".txt")) ||
+          path.EndsWithIgnoreCase(_T(".air")) ||
+          path.EndsWithIgnoreCase(_T(".sua"))))
+      continue;
```
This makes this only usable for these types of files. What if for example we would like to do waypoints?

Eldalie replied:
Yes, this PR focuses on airspace only; the next one will address multiple waypoints.
